urlfaa9045c2e_Azmisov_racko_ai_tgz-pJ8-client_TestingJ8/src/models/DataInstance.java

```diff
@@ -0,0 +1,58 @@
+package models;
+
+/**
+ * Generic inputs instance for neural network training
+ */
+public class DataInstance {
+	public double[] inputs;
+	public double output;
+	private int loc = 0;
+	
+	public DataInstance(int features){
+		inputs = new double[features];
+	}
+	
+	//Integer features
+	public void addFeature(int[] vals, double normalize){
+		for (int i=0; i<vals.length; i++)
+			inputs[loc++] = vals[i] / normalize;
+	}
+	public void addFeature(int val, double normalize){
+		inputs[loc++] = val / normalize;
+	}
+	
+	//Double features
+	public void addFeature(double[] vals, double normalize){
+		if (normalize == 1){
+			System.arraycopy(vals, 0, inputs, loc, vals.length);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;vals&#39; (index: 0 | type: double[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;normalize&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).